simplify module argspec vs doc type mismatch checks, display - by default

bcoca opened the pull request:
simplify checks for type on modules, only care if they do not match
fixes #58645
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
docs/tests

bcoca committed:
display str as type default for string

  simplify checks for type on modules, only care if they do not match

bcoca committed:
removed docs from removed tests

felixfontein commented on test/sanity/validate-modules/main.py:
```diff
@@ -1251,39 +1251,21 @@ def _validate_argument_spec(self, docs, spec, kwargs):
+            doc_type = docs.get('options', {}).get(arg, {}).get('type', 'str')
+            arg_type = data.get('type', 'str')
+            if doc_type is None:
+                doc_type == 'str'
```
```suggestion
                doc_type = 'str'
```

felixfontein commented on test/sanity/validate-modules/main.py:
```diff
@@ -1251,39 +1251,21 @@ def _validate_argument_spec(self, docs, spec, kwargs):
+            if doc_type is None:
+                doc_type == 'str'
+            if arg_type is None:
+                arg_type == 'str'
```
```suggestion
                arg_type = 'str'
```

bcoca committed:
Update test/sanity/validate-modules/main.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

felixfontein commented on test/sanity/validate-modules/main.py:
```diff
@@ -1251,39 +1251,21 @@ def _validate_argument_spec(self, docs, spec, kwargs):
                          "but documentation defines default as (%r)" % (arg, arg_default, doc_default))
                 )
 
+
```
```suggestion
```

bcoca committed:
we always have arg_type

bcoca committed:
whitesnake

Update ignore.txt.

remove dupes

updated as per meeting, default to -, not string

remove dupes

removed

now pass

felixfontein commented:
I don't know if such a script already exists, but here's one which updates `test/sanity/validate-modules/ignore.txt` based on the `test/runner/ansible-test sanity --test validate-modules --docker-no-pull --docker` output: https://gist.github.com/felixfontein/219b19f8c39b7755b4482e3c28212d7e

dagwieers commented on hacking/build_library/build_ansible/command_plugins/plugin_formatter.py:
```diff
@@ -153,16 +153,17 @@ def rst_xline(width, char="="):
+    # list/path/etc are already the same as their description
+    # we only put here those that need 'long name translation'.
+    if text == 'str' or isinstance(text, Undefined):
+        text = 'string'
```
We don't want to say an option is a **string** if that is not certain. Most modules have no type-information for the docs, so that means suddenly all parameters become strings.

That is the reason why I proposed it as part of: https://github.com/ansible/ansible/pull/45805

But only part of that PR was accepted, which meant that if you wanted type-information in the docs, you have to provide it separately. (Which I started doing for all my modules, the Windows modules and a large part of system modules, but stopped after my PRs became stale)

In any case, this change is wrong, and *even* if we could use the argspec in the future, there will always be a need to document the type differently then the argspec (i.e. in case of raw for backward compatibility).

dagwieers replied:
Also see: https://github.com/ansible/ansible/pull/56312

felixfontein commented:
(Removed conflicts, updated `ignore.txt`)

felixfontein commented:
(Removed conflicts, updated `ignore.txt`)

felixfontein committed:
Add changelog.

felixfontein committed:
Update ignore.txt.

felixfontein commented on test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py:
```diff
@@ -1358,50 +1358,22 @@ def _validate_argument_spec(self, docs, spec, kwargs, context=None):
+            doc_type = doc_options_arg.get('type', 'str')
+            arg_type = data.get('type', 'str')
+            if doc_type is None:
+                doc_type = 'str'
```
I'm not sure whether we should really do that. There are still too many modules which don't declare their types, so as a user (with module development background) I'm way more happy if `str` is explicitly mentioned in the documentation.

felixfontein commented:
(Tests are currently failling because AIX instances aren't starting. Should be completely unrelated to this PR.)

samccann reviewed (approved):
docs portion LGTM
